src/main/java/com/mparticle/kits/ResponsysKit.java

```diff
@@ -0,0 +1,304 @@
+                Map<String, Object> eventProperties = new HashMap<>();
+                eventProperties.put("Pid", product.getSku());
+                eventProperties.put("Pc", product.getCategory());
+
+                mPushIOManager.trackEvent(responsysEvent, eventProperties);
+
+                if (productAction.equalsIgnoreCase(Product.PURCHASE)) {
+                    mPushIOManager.trackEngagement(PushIOManager.PUSHIO_ENGAGEMENT_METRIC_PURCHASE);
+                }
+            }
+
+            List<ReportingMessage> reportingMessages = new ArrayList<>();
+            reportingMessages.add(ReportingMessage.fromEvent(this, commerceEvent));
+            return reportingMessages;
+        }
+
+        return null;
+    }
+
+    @Override
+    public List<ReportingMessage> leaveBreadcrumb(String s) {
+        return null;
+    }
+
+    @Override
+    public List<ReportingMessage> logError(String s, Map<String, String> map) {
+        return null;
+    }
+
+    @Override
+    public List<ReportingMessage> logException(Exception e, Map<String, String> map, String s) {
+        return null;
+    }
+
+    @Override
+    public List<ReportingMessage> logEvent(MPEvent mpEvent) {
+        if (mpEvent == null) {
+            return null;
+        }
+
+        List<ReportingMessage> reportingMessages = new ArrayList<>();
+
+        final Map<String, List<String>> customFlags = mpEvent.getCustomFlags();
+        final EventType eventType = mpEvent.getEventType();
+
+        PIOLogger.v("RK lE event type: " + eventType);
+
+        switch (eventType) {
+            case Search:
+                mPushIOManager.trackEvent("$Searched");
+                reportingMessages.add(ReportingMessage.fromEvent(this, mpEvent));
+                break;
+            case Social:
+                mPushIOManager.trackEngagement(PushIOManager.PUSHIO_ENGAGEMENT_METRIC_SOCIAL);
+                reportingMessages.add(ReportingMessage.fromEvent(this, mpEvent));
+                break;
+            case UserPreference:
+                if (customFlags != null && customFlags.containsKey(CUSTOM_FLAG_PREFS)) {
+                    Map<String, String> eventInfo = mpEvent.getInfo();
+                    if (eventInfo != null) {
+                        for (Map.Entry<String, String> entry : eventInfo.entrySet()) {
+                            try {
+                                mPushIOManager.declarePreference(entry.getKey(), entry.getKey(), PushIOPreference.Type.STRING);
+                                mPushIOManager.setPreference(entry.getKey(), entry.getValue());
+                            } catch (ValidationException e) {
+                                PIOLogger.v("RK lE ValidationException: " + e.getMessage());
+                            }
+                        }
+                    }
+                }
+                break;
+            case Other:
+                if (customFlags != null && customFlags.containsKey(CUSTOM_FLAG_IAM)) {
```

Comment:
So is this the only case where the integration will support custom event names - if they customer has set a custom flag? That doesn't seem ideal - why not just pass all events through with their customer-given name as well as their properties map (which i noticed isn't passed anywhere here except for preferences above)?

My concern with the current implementation is that customers are going to get *a lot* of "trackEngagement" calls that are not going to be discernible from each other.

Also - this implementation is very different than the iOS kit PR. Please make the two identical.

Reply:
Please review commit: e4dd26c1702468166e1b3a55e717bf53a6e30637